snowmelt_timing/snowmelt_timing.R

```diff
@@ -0,0 +1,142 @@
+
+# pull 2021 water year SWE up to current date
+get_SWE_2021 <- function(site, date_today) {
+  
+  date <- as.Date(date_today)
+  site_meta <- sntl_meta %>% filter(site_id == site)
+  year <- word(site_meta$start, 1,1, sep="-")
+  sitey <- gsub('SNTL:', '', site_meta$site_id)
+  
+  ## pull data for nrcs historical range
+  sntl <- grabNRCS.data(network = 'SNTL', 
+                        site_id = sitey, 
+                        timescale = 'daily', 
+                        DayBgn = sprintf("%s-10-01", year),
+                        DayEnd = date)
+  print(i)
+  
+  if("Snow.Water.Equivalent..in..Start.of.Day.Values" %in% colnames(sntl)){
+    
+    sntl_clean <- clean_SWE(today = sntl) %>% mutate(site_id = sitey)
+    write.csv(sntl_clean, sprintf('swe_%s.csv', sitey),row.names=FALSE)
+    return(sntl_clean)
+    
+  }
+}
+
+clean_SWE <- function(today){
+  
+  # organize data
+  site_curve <- today %>%
+    mutate(date = as.Date(Date), 
+           year = year(date), month = month(date), 
+           water_year = year(date) + ifelse(month(date) >= 10, 1, 0), 
+           water_day = (date - as.Date(sprintf('%s-10-01', water_year)))) %>%
+    group_by(water_year) %>%
+    mutate(water_day_max = min(water_day)) %>% ungroup() %>%
+    mutate(water_day = as.numeric(water_day - water_day_max + 1)) %>%
+    ungroup() %>%
+    mutate(swe = Snow.Water.Equivalent..in..Start.of.Day.Values) %>%
+    dplyr::select(date, water_year, water_day, swe, year) 
+  
+  return(site_curve)
+}
+
+## pull SNOTEL data for all currently active sites from inception to today
+site_list <- unique(sntl_meta$site_id)
+
+## this takes A WHILE!
+for (i in 1:length(site_list)) {
+  swe_out <- tryCatch(get_SWE_2021(site=site_list[i], date_today = Sys.Date()),
+                      error = function(e) return())
+}
+
+
+# calculate snowmelt timing -----------------------------------------------
+
+# read in data
+swe_files <- list.files(pattern="swe", full.names=TRUE)
+
+swe_data <- lapply(swe_files, read_csv) %>% 
+  bind_rows() %>%
+  filter(water_year >= 1981 & water_year <= 2020)
+
+
+# find 50% of peak swe for each year and site
+peaks <- swe_data %>%
+  group_by(site_id, water_year)%>%
+  summarize(peak_swe = max(swe, na.rm=TRUE)) %>%
+  mutate(sm50_swe = peak_swe/2)
+
+# find date of peak swe each year
+peak_dates <- swe_data %>%
+  left_join(peaks) %>%
+  group_by(site_id, water_year)%>%
+  filter(swe == peak_swe)%>%
+  filter(water_day == max(water_day, na.rm=TRUE))%>% # take latest date peak occurs
+  mutate(peak_day = water_day)%>%
+  dplyr::select(-date, -water_day, -swe, -year) %>%
+  ungroup()
+
+# find date of sm50_swe for each year
+sm50_dates <- swe_data %>%
+  left_join(peak_dates) %>%
+  group_by(site_id, water_year)%>%
+  filter(water_day >= peak_day) %>%
+  filter(swe <= sm50_swe) %>%
+  filter(water_day == min(water_day, na.rm=TRUE)) %>% # earliest water day at or below peak/2
```

Comment:
Are these the methods that the snow folks use? Just curious since peak uses latest day and SM50 uses earliest.

Reply:
Yeah, I did peak SWE this way because SNOTEL uses the latest date at the maximum SWE for their metric `Snow Water Equivalent Date of Water Year Peak` https://www.nrcs.usda.gov/wps/wcm/connect/wcc/1eb4ef2b-b35b-447a-a177-eed8968b2787/iMap_Glossary.pdf?MOD=AJPERES&CVID=noqKkGW&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mrGiEis&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp&CVID=mr5UxEp

Reply:
and for the melt date, it's the first day that SWE is at or lower than 50%  of that max